File: app/src/main/java/com/thatcakeid/zrytezene/adapters/HomeItemsRecyclerViewAdapter.java
package com.thatcakeid.zrytezene.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.google.firebase.Timestamp;
import com.thatcakeid.zrytezene.R;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;

public class HomeItemsRecyclerViewAdapter extends RecyclerView.Adapter<HomeItemsRecyclerViewAdapter.ViewHolder> {

    private ArrayList<HashMap<String, Object>> items;
    private HashMap<String, String> users;
    private Context mContext;

    private static ClickListener clickListener;

    public HomeItemsRecyclerViewAdapter(Context mContext, ArrayList<HashMap<String, Object>> items,
                                        HashMap<String, String> users) {
        this.mContext = mContext;
        this.items = items;
        this.users = users;
    }

    public void updateItems(Context mContext, ArrayList<HashMap<String, Object>> items,
                            HashMap<String, String> users) {
        this.mContext = mContext;
        this.items = items;
        this.users = users;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.home_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.musicname_item.setText(items.get(position).get("title").toString());
        holder.uploader_item.setText(users.containsKey(items.get(position).get("author").toString()) ?
                users.get(items.get(position).get("author").toString()) :
                items.get(position).get("author").toString());
        holder.date_text.setText(new SimpleDateFormat("EEE, d MMM yyyy HH:mm:ss")
                .format(((Timestamp)items.get(position).get("time")).toDate()));
        if (items.get(position).get("thumb").toString().equals("")) {
            holder.image_overlay.setImageResource(R.drawable.ic_zrytezene);
        } else {
             Glide.with(mContext).load(items.get(position).get("thumb").toString())
                     .into(holder.image_overlay);
        }
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {

        ImageView image_overlay;
        TextView musicname_item;
        TextView uploader_item;
        TextView date_text;

        public ViewHolder(View itemView) {
            super(itemView);

            itemView.setOnClickListener(this);
            itemView.setOnLongClickListener(this);

            image_overlay = itemView.findViewById(R.id.image_overlay);
            musicname_item = itemView.findViewById(R.id.musicname_item);
            uploader_item = itemView.findViewById(R.id.uploader_item);
            date_text = itemView.findViewById(R.id.date_text);
        }

        @Override
        public void onClick(View v) {
            clickListener.onItemClick(getAdapterPosition(), v);
        }

        @Override
        public boolean onLongClick(View v) {
            clickListener.onItemLongClick(getAdapterPosition(), v);
            return false;
        }
    }

    public void setOnItemClickListener(ClickListener clickListener) {
        HomeItemsRecyclerViewAdapter.clickListener = clickListener;
    }

    public interface ClickListener {
        void onItemClick(int position, View v);
        void onItemLongClick(int position, View v);
    }
}

File: app/src/main/java/com/thatcakeid/zrytezene/adapters/HomeItemsRecyclerViewAdapter.kt
package com.thatcakeid.zrytezene.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.thatcakeid.zrytezene.HelperClass.Companion.getPrettyDateFormat
import com.thatcakeid.zrytezene.HelperClass.Companion.getPrettyPlaysCount
import com.thatcakeid.zrytezene.R
import com.thatcakeid.zrytezene.data.MusicEntry

class HomeItemsRecyclerViewAdapter(
    private var mContext: Context,
    private var items: List<MusicEntry>,
    private var users: Map<String, String>
) : RecyclerView.Adapter<HomeItemsRecyclerViewAdapter.ViewHolder>() {

    fun updateItems(
        mContext: Context,
        items: List<MusicEntry>,
        users: Map<String, String>
    ) {
        this.mContext = mContext.applicationContext
        this.items = items
        this.users = users
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.home_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val bottomText = (
                if (users.containsKey(items[position].authorUserId)) users[items[position].authorUserId]
                else items[position].authorUserId).toString() + " • " +

                getPrettyPlaysCount(items[position].plays) + " • " +
                getPrettyDateFormat(items[position].time)

        holder.musicNameItem.text = items[position].title
        holder.uploaderItem.text = bottomText

        if (items[position].thumb == null) {
            holder.imageOverlay.setImageResource(R.drawable.ic_zrytezene)
        } else {
            Glide.with(mContext)
                .load(items[position].thumb)
                .into(holder.imageOverlay)
        }
    }

    override fun getItemCount(): Int = items.size

    class ViewHolder(itemView: View)
        : RecyclerView.ViewHolder(itemView), View.OnClickListener, OnLongClickListener {

        var imageOverlay: ImageView = itemView.findViewById(R.id.image_overlay)
        var musicNameItem: TextView = itemView.findViewById(R.id.musicname_item)
        var uploaderItem: TextView = itemView.findViewById(R.id.uploader_item)

        override fun onClick(v: View) {
            clickListener?.onItemClick(adapterPosition, v)
        }

        override fun onLongClick(v: View): Boolean {
            clickListener?.onItemLongClick(adapterPosition, v)
            return false
        }

        init {
            itemView.setOnClickListener(this)
            itemView.setOnLongClickListener(this)
        }
    }

    fun setOnItemClickListener(clickListener: ClickListener) {
        Companion.clickListener = clickListener
    }

    interface ClickListener {
        fun onItemClick(position: Int, v: View)
        fun onItemLongClick(position: Int, v: View)
    }

    companion object {
        private var clickListener: ClickListener? = null
    }
}